### PR TITLE
improve e2e retry logic with standard wait.Poll()

### PR DESCRIPTION
**Pull request by dalanlan**
improve original `for` retry logic with standard `wait.Poll` in e2e tests.
Signed-off-by: He Simei hesimei@zju.edu.cn

### CODE REVIEW

**Comment by dalanlan** [comment]
@quinton-hoole

File: test/e2e/es_cluster_logging.go [hunk 1]
```diff
@@ -72,12 +73,14 @@ func ClusterLevelLoggingWithElasticsearch(c *client.Client) {
 	// being run as the first e2e test just after the e2e cluster has been created.
 	var err error
 	const graceTime = 10 * time.Minute
-	for start := time.Now(); time.Since(start) < graceTime; time.Sleep(5 * time.Second) {
+	start := time.Now()
```
Comment (quinton-hoole) [inline test/e2e/es_cluster_logging.go]: Can remove this and use graceTime instead in logging.

Comment (quinton-hoole) [inline test/e2e/es_cluster_logging.go]: Sorry - I'm confused.  Ignore the above comment.

File: test/e2e/es_cluster_logging.go [hunk 2]
```diff
@@ -72,12 +73,14 @@ func ClusterLevelLoggingWithElasticsearch(c *client.Client) {
...
+	expectNoError(wait.Poll(5*time.Second, graceTime, func() (bool, error) {
 		if _, err = s.Get("elasticsearch-logging"); err == nil {
-			break
+			return true, nil
 		}
 		Logf("Attempt to check for the existence of the Elasticsearch service failed after %v", time.Since(start))
-	}
+		return false, nil
```
Comment (quinton-hoole) [inline test/e2e/es_cluster_logging.go]: Ideally you should return (true, err) here in cases of errors which are not worth retrying. That way you fail earlier, and get a more useful error message as output (e.g. "malformed request x" rather then "timed out").

File: test/e2e/kubectl.go [hunk 3]
```diff
@@ -144,13 +145,20 @@ func validateGuestbookApp(c *client.Client, ns string) {
 
 // Returns whether received expected response from guestbook on time.
 func waitForGuestbookResponse(c *client.Client, cmd, arg, expectedResponse string, timeout time.Duration, ns string) bool {
-	for start := time.Now(); time.Since(start) < timeout; time.Sleep(5 * time.Second) {
+	flag := 1
```
Comment (quinton-hoole) [inline test/e2e/kubectl.go]: This needs a better name.

Comment (quinton-hoole) [inline test/e2e/kubectl.go]: And it should be a boolean, not an int, right?

File: test/e2e/kubectl.go [hunk 4]
```diff
@@ -144,13 +145,20 @@ func validateGuestbookApp(c *client.Client, ns string) {
...
+	expectNoError(wait.Poll(5*time.Second, timeout, func() (bool, error) {
 		res, err := makeRequestToGuestbook(c, cmd, arg, ns)
 		if err == nil && res == expectedResponse {
-			return true
+			flag = 0
+			return true, nil
 		}
+		return false, nil
+	}))
+	if flag == 0 {
```
Comment (quinton-hoole) [inline test/e2e/kubectl.go]: Can you actually get here without flag being zero?  If flag is 1, then you never succeeded, in which case Poll returns an error, expectNoError fails etc.

File: test/e2e/pods.go [hunk 5]
```diff
@@ -98,21 +98,27 @@ func testHostIP(c *client.Client, pod *api.Pod) {
 	err = waitForPodRunningInNamespace(c, pod.Name, ns)
 	Expect(err).NotTo(HaveOccurred())
 	// Try to make sure we get a hostIP for each pod.
-	hostIPTimeout := 2 * time.Minute
-	t := time.Now()
-	for {
-		p, err := podClient.Get(pod.Name)
+
+	var (
+		flag          = 1
```
Comment (quinton-hoole) [inline test/e2e/pods.go]: Similar comments to above.  t, p and flag need more descriptive variable names.

File: test/e2e/reboot.go [hunk 6]
```diff
@@ -270,9 +272,15 @@ func waitForNodeToBe(c *client.Client, name string, wantReady bool, timeout time
...
 				Logf("Successfully found node %s readiness to be %t", name, wantReady)
-				return true
+				flag = 0
+				return true, nil
 			}
+
 		}
+		return false, nil
+	}))
+	if flag == 0 {
```
Comment (quinton-hoole) [inline test/e2e/reboot.go]: Similar comment to above. wait.Poll() returns an error if it failed, so you don't need this flag.

File: test/e2e/service.go [hunk 7]
```diff
@@ -245,11 +242,7 @@ var _ = Describe("Services", func() {
 
 		validateEndpointsOrFail(c, ns, serviceName, map[string][]int{})
 
-		// We deferred Gingko pieces that may Fail, we aren't done.
-		defer func() {
-			close(done)
-		}()
-	}, 240.0)
```
Comment (quinton-hoole) [inline test/e2e/service.go]: Why did you remove this block?

File: test/e2e/util.go [hunk 8]
```diff
@@ -389,16 +391,15 @@ func testContainerOutputInNamespace(scenarioName string, c *client.Client, pod *
 			Suffix("containerLogs", ns, podStatus.Name, containerName).
 			Do().
 			Raw()
-		fmt.Sprintf("pod logs:%v\n", string(logs))
-		By(fmt.Sprintf("pod logs:%v\n", string(logs)))
+		Logf("pod logs:%v\n", string(logs))
+		Logf("pod logs:%v\n", string(logs))
```
Comment (quinton-hoole) [inline test/e2e/util.go]: Why twice?

**Comment by quinton-hoole** [comment]
Thanks very much for this cleanup Emma.  I provided a few comments, most of which apply to multiple tests, but I only commented on the first occurrence.   Let me know when you want me to re-review.

File: test/e2e/service.go (hunk 7 above)
Comment (dalanlan) [inline test/e2e/service.go]: Move the retry logic to `validateEndpointsOrFail` instead.

File: test/e2e/es_cluster_logging.go (hunk 2 above)
Comment (dalanlan) [inline test/e2e/es_cluster_logging.go]: Some test cases are worth retrying, right? 
I wonder how you distinguish them:)

**Comment by dalanlan** [comment]
Updated. @quinton-hoole  
I do not change the behavior of retry block (`return true, err` stuff) for I think it is sort of hard to tell apart cases worth retrying or not. 
What's more, I guess the comment here(https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/es_cluster_logging.go#L71) also indicates that retry should be conducted. 
Let me know if you insist.

**Comment by quinton-hoole** [comment]
LGTM!  Thanks again for this tidy up Emma!

**Comment by satnam6502** [comment]
This PR has broken the cluster level logging e2e test. Please can it be reverted.

**Comment by satnam6502** [comment]
I don't see how this PR could have passed an e2e test? Surely one should have been run before merge?

**Comment by quinton-hoole** [comment]
My mistake @satnam6502 . I had implicitly assumed that the submitter had successfully run the e2e test suite, but forgot to check this explicitly.  Rolling back now.

**Comment by satnam6502** [comment]
If the roll-back is too painful I could offer to submit a new PR that puts the e2e logging test back to its original form.

**Comment by satnam6502** [comment]
The root of the issue is that when the test passed, it indicates this with a `return nil...` statement. This is now in a closure which means it will just return and resume executing in the enclosing function which has a final statement which always indicates failure. Such `return` statements should never be embedded in closures.

**Comment by dalanlan** [comment]
My bad. i didnt run it on es_clustering_logging for i am using ubuntu.

Sorry but i could not quite catch you. What do you say about this? 
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/pods.go#L295

**Comment by satnam6502** [comment]
I am utterly swamped with other things and I can't find the time to look at this util we are > v1.0
